WordPress/gutenberg
# Introduce a common API to style Button from both web and mobile

## Description
This is an exploration of x-platform style system. https://github.com/wordpress-mobile/gutenberg-mobile/issues/1411
This is the second iteration on https://github.com/wordpress-mobile/gutenberg-mobile/issues/1386

Mobile side PR: https://github.com/WordPress/gutenberg/pull/19179

In this PR I used `styled-system` with `emotion` to create a primitive Button component that could be used in web and mobile and could be styled via props.

- I added `A` and `button` components since both of them are used in `@components/Button`. They accept `className` as a prop to be backward compatibility ( we need to support already existing themes ) and all style props that are supported by `styled-system`
- I added a `PrimitiveButton` that is described here - https://github.com/WordPress/gutenberg/pull/19104#issuecomment-580780725
- I added the theme provider and theme values (they could be a bit outdated since I took them from my previous PR https://github.com/WordPress/gutenberg/pull/17614 )
- I used created primitives in `@components/Button` (after that all buttons uses the primitives)
- I passed styles via props instead of CSS in the Inserter button.
- In addition, I created `BlockBreadcrumbButton` components that use PrimitiveButton instead of `@components/Button` just to demonstrate.

## How has this been tested?
- Run Gutenberg web
- All buttons should look the same as before
- Check if `BlockBreadcrumbButton` looks correct with and without hover/focus
- Check if Inserter button looks correct
- The preview button should be different than originally - it should be 50px height and has bigger borderRadius.

## Screenshots 
The `BreadcrumbButton` component that uses `PrimitiveButton` with `additionalStyles` prop to add `hover`, `focused` styles - https://github.com/WordPress/gutenberg/pull/19104/files#diff-21bd8baf92167b81915be200a8dd9eb8
<img width="1001" alt="Screenshot 2019-12-17 at 12 59 43" src="https://user-images.githubusercontent.com/16336501/70993993-80d44180-20cd-11ea-8370-59638be4c0a1.png">
### :hover
<img width="1078" alt="Screenshot 2019-12-17 at 12 59 33" src="https://user-images.githubusercontent.com/16336501/70993984-7dd95100-20cd-11ea-9a4d-a0b6ad84793a.png">

### :hover:focus
<img width="545" alt="Screenshot 2020-02-10 at 15 57 35" src="https://user-images.githubusercontent.com/16336501/74161082-b05f8000-4c1e-11ea-9a7a-3e778e8eb028.png">

### :hover:focus:active
<img width="651" alt="Screenshot 2020-02-10 at 15 57 50" src="https://user-images.githubusercontent.com/16336501/74161055-a89fdb80-4c1e-11ea-85fd-505d96b42e4b.png">


The `Inserter` button that uses `@components/Button` with some changed styles by `additionalStyles` and `styled-system` props - https://github.com/WordPress/gutenberg/pull/19104/files#diff-c6139f159c9b2b79af303587407aea51

<img width="673" alt="Screenshot 2019-12-17 at 13 01 47" src="https://user-images.githubusercontent.com/16336501/70993999-8598f580-20cd-11ea-8b9f-7db32ae06ade.png">

### :hover
<img width="824" alt="Screenshot 2020-02-10 at 15 58 07" src="https://user-images.githubusercontent.com/16336501/74162633-5b713900-4c21-11ea-87d6-8348faca91e1.png">

### :hover:focus
<img width="700" alt="Screenshot 2020-02-10 at 15 58 41" src="https://user-images.githubusercontent.com/16336501/74162667-6af08200-4c21-11ea-8a7d-217b26a9e47c.png">

### :hover:focus:active
<img width="791" alt="Screenshot 2020-02-10 at 15 59 02" src="https://user-images.githubusercontent.com/16336501/74162707-7f347f00-4c21-11ea-978a-02544226be7f.png">


## Types of changes
Experimenting with `emotion` and `styled-system`

## PrimitiveButton checklist
- [x] States
- [x] States change
  - on the web, we can use `onFocus`, `onBlur` to determine the `focus` state. MouseEvents to determine the `active` and `hover` state. We specify whether the button is enabled or disabled by prop so we don't need to listen to that change. The only state left is `visited` but I'm not sure if we can get that information somehow and also if the `visited` is also applicable to button or only to the `a` element.
- [x] Icons & Labels
- [x] Classname
- [x] Visibility
- [x] CSS prop

## components/Button checklist
- [x] Components/Button should have the same props PrimitiveButton has + variants
- [x] Styles passed via props will override variant styles
  - They do but i.e even if i use `marginLeft` in the `inserter` button, this is not applied because of styles that are declared in a different place and are more specific (i think it should work like that because of themes that can override styles) but they override styles that are declared in `components/Button`
- [x]  keep the backward compatibility

## Checklist:
